examples/vision/learnable_resizer.py

```diff
@@ -0,0 +1,273 @@
+resizing methods like **bilinear interpolation** for this step and the resized images do
+not lose much of their perceptual character to the human eyes. In [Learning to Resize
+Images for Computer Vision Tasks](https://arxiv.org/abs/2103.09950v1), Talebi et. al show
+that if we try to optimize the perceptual quality of the images for the vision models
+rather than the human eyes, their performance can further be improved. They investigate
+the following question: 
+
+**For a given image resolution and a model, how to best resize the given images?**
+
+As shown in the paper, this idea helps to consistently improve the performance of the
+common vision models (pre-trained on ImageNet-1k) like DenseNet121, ResNet50,
+MobileNetV2, and EfficientNets. In this example, we will implement the learnable image
+resizing module as proposed in the paper and demonstrate that on the
+[Cast-Vs-Dogs dataset](https://www.microsoft.com/en-us/download/details.aspx?id=54765)
+using the [DenseNet121](https://arxiv.org/abs/1608.06993) model. 
+
+This example requires TensorFlow 2.4 or higher.
+"""
+
+"""
+## Setup
+"""
+
+from tensorflow.keras import layers
+from tensorflow import keras
+import tensorflow as tf
+
+import tensorflow_datasets as tfds
+
+tfds.disable_progress_bar()
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+"""
+## Define hyperparameters
+"""
+
+"""
+In order to facilitate mini-batch learning, we need to have a fixed shape for the images
+inside a given batch. This is why an initial resizing is required. We first resize all
+the images to (300x300) shape and then learn their optimal representation for the
+(224x224) resolution. 
+"""
+
+INP_DIM = (300, 300)
+TARGET_DIM = (224, 224)
+INTERPOLATION = "bilinear"
+
+AUTO = tf.data.AUTOTUNE
+BATCH_SIZE = 64
+EPOCHS = 5
+ALPHA = 0.2
+
+"""
+For this example, we will use the bilinear interpolation but the learnable image resizer
+module is not dependent on any specific interpolation method. We can use other ones (such
+as bicubic) as well. 
+"""
+
+"""
+## Load and prepare the dataset
+
+For this example, we will only use 40% of the total training dataset.
+"""
+
+train_ds, validation_ds = tfds.load(
+    "cats_vs_dogs",
+    # Reserve 10% for validation
+    split=["train[:40%]", "train[40%:50%]"],
+    as_supervised=True,
+)
+
+
+def preprocess_dataset(image, label):
+    image = tf.image.resize(image, (INP_DIM[0], INP_DIM[1]))
+    label = tf.one_hot(label, depth=2)
+    return (image, label)
+
+
+train_ds = (
+    train_ds.shuffle(BATCH_SIZE * 100)
+    .map(preprocess_dataset, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+validation_ds = (
+    validation_ds.map(preprocess_dataset, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+"""
+## Define the learnable resizer utilities
+
+The figure below (courtesy of the original paper) presents the structure of the learnable
+resizing module:
```

Comment:
Suggestion:

- There are at least 2 papers here, so to make it clear which "original paper" this was taken from, let's be more specific: it's from Learning to Resize Images for Computer Vision Tasks by Talebi & Milanfar (2021)


```suggestion
Below is the structure of the learnable resizing module (source: Learning to
Resize Images for Computer Vision Tasks by Talebi & Milanfar (2021)):
```